crates/libtest2/src/macros.rs

```diff
@@ -99,3 +121,80 @@ macro_rules! _parse_ignore {
+        run($context)
+    }};
+}
+
+#[macro_export]
+macro_rules! assert_panic {
+    ($f:expr, $expected:literal $(,)?) => {
+        match ::std::panic::catch_unwind(::std::panic::AssertUnwindSafe(|| $f)) {
+            // The test should have panicked, but didn't.
+            ::std::result::Result::Ok(_) => {
+                // TODO: Rust includes the source file location here, consider doing the same?
+                ::std::result::Result::Err($crate::RunError::fail("test did not panic as expected"))
+            }
+
+            // The test panicked, as expected.
+            ::std::result::Result::Err(payload) => {
+                // The `panic` information is just an `Any` object representing the
+                // value the panic was invoked with. For most panics (which use
+                // `panic!` like `println!`), this is either `&str` or `String`.
+                let maybe_panic_str = payload
+                    .downcast_ref::<::std::string::String>()
+                    .map(|s| s.as_str())
+                    .or_else(|| payload.downcast_ref::<&str>().copied());
+
+                // Enforce `$expected` to be a string literal
+                let expected: &'static str = $expected;
+
+                // Check the panic message against the expected message.
+                match maybe_panic_str {
+                    ::std::option::Option::Some(panic_str) if panic_str.contains(expected) => {
+                        ::std::result::Result::Ok(())
+                    }
+
+                    ::std::option::Option::Some(panic_str) => {
+                        let error_msg = ::std::format!(
+                            r#"panic did not contain expected string
+      panic message: {panic_str:?}
+ expected substring: {expected:?}"#
+                        );
+
+                        ::std::result::Result::Err($crate::RunError::fail(error_msg))
+                    }
+
+                    ::std::option::Option::None => {
+                        let type_id = (*payload).type_id();
+                        let error_msg = ::std::format!(
+                            r#"expected panic with string value,
+ found non-string value: `{type_id:?}`
+     expected substring: {expected:?}"#,
+                        );
+
+                        ::std::result::Result::Err($crate::RunError::fail(error_msg))
+                    }
+                }
+            }
+        }
+    };
+    ($f:expr $(,)?) => {
+        match ::std::panic::catch_unwind(::std::panic::AssertUnwindSafe(|| $f)) {
+            // The test should have panicked, but didn't.
+            ::std::result::Result::Ok(_) => {
+                // TODO: Rust includes the source file location here, consider doing the same?
+                ::std::result::Result::Err($crate::RunError::fail("test did not panic as expected"))
+            }
+
+            // The test panicked, as expected.
+            ::std::result::Result::Err(_) => Ok(()),
+        }
```

Comment:
Would there be a good way to unify these code paths, like turning `expected` into an `Option<&str`>?